stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,255 @@
+        index ^= index >> (_Wait_table_size_power * 2);
+        index ^= index >> _Wait_table_size_power;
+        return wait_table[index & _Wait_table_index_mask];
+    }
+
+    static constexpr std::size_t _Uninitialized_spin_count = (std::numeric_limits<std::size_t>::max)();
+    static std::atomic<std::size_t> _Atomic_spin_count     = _Uninitialized_spin_count;
+
+    std::size_t _Atomic_init_spin_count() noexcept {
+        std::size_t result = (std::thread::hardware_concurrency() == 1 ? 0 : 10'000) * _Atomic_spin_value_step;
+        _Atomic_spin_count.store(result, std::memory_order_relaxed);
+        // Make sure other thread is likely to get this,
+        // as we've done kernel call for that.
+        std::atomic_thread_fence(std::memory_order_seq_cst);
+        return result;
+    }
+
+
+#if _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8
+    constexpr bool _Have_wait_functions() {
+        return true;
+    }
+#define __crtWaitOnAddress       WaitOnAddress
+#define __crtWakeByAddressSingle WakeByAddressSingle
+#define __crtWakeByAddressAll    WakeByAddressAll
+
+#pragma comment(lib, "Synchronization.lib")
+
+    void _Atomic_wait_fallback(
+        [[maybe_unused]] const void* const _Storage, [[maybe_unused]] unsigned long long& _Wait_context) noexcept {
+        std::terminate();
+    }
+
+    void _Atomic_notify_fallback([[maybe_unused]] const void* const _Storage) noexcept {
+        std::terminate();
+    }
+
+    void _Atomic_unwait_fallback(
+        [[maybe_unused]] const void* const _Storage, [[maybe_unused]] unsigned long long& _Wait_context) noexcept {}
+
+#else // ^^^ _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8 / _STL_WIN32_WINNT < _WIN32_WINNT_WIN8 vvv
+    void _Atomic_wait_fallback(const void* const _Storage, unsigned long long& _Wait_context) noexcept {
+        switch (_Wait_context & _Atomic_wait_phase_mask) {
+        case _Atomic_wait_phase_wait_none: {
+            _Wait_context = _Atomic_wait_phase_wait_locked;
+            auto& entry   = _Atomic_wait_table_entry(_Storage);
+            ::AcquireSRWLockExclusive(&entry._Lock);
+            [[fallthrough]];
+        }
+
+        case _Atomic_wait_phase_wait_locked: {
+            auto& entry = _Atomic_wait_table_entry(_Storage);
+            ::SleepConditionVariableSRW(&entry._Condition, &entry._Lock, INFINITE, 0);
+            // re-check, and still in _Atomic_wait_phase_wait_locked
+            return;
+        }
+        }
+    }
+
+    void _Atomic_unwait_fallback(const void* const _Storage, unsigned long long& _Wait_context) noexcept {
+        if (_Wait_context & _Atomic_wait_phase_wait_locked) {
+            auto& entry = _Atomic_wait_table_entry(_Storage);
+            ::ReleaseSRWLockExclusive(&entry._Lock);
+        }
+    }
+
+    void _Atomic_notify_fallback(const void* const _Storage) noexcept {
+        auto& entry = _Atomic_wait_table_entry(_Storage);
+        ::AcquireSRWLockExclusive(&entry._Lock);

```

Comment:
Why acquire and then immediately release? I'm almost certain that the notify needs to be done while holding the mutex to prevent missed wakes in this pattern:

```
T1:
unique_lock lck(mtx);
for (;;) {
auto local = load();
if (42 != local) { return; }
cv.wait(lck);
}

T2:
store(42);
cv.notify_all();
```

to prevent the following deadlock:

```
T1: take lock
T1: load local
T1: see that it isn't 42
T2: store(42)
T2: notify_all()
T1: wait() // deadlock
```

(If this deadlock is impossible there should be a comment here describing why)

Reply:
> T1: see that it isn't 42

I take it as "T1: see that it is still 42", otherwise we'll just return.

In this sequence, when T2 calls `notify_all()`, locking on a mutex will postpone actual notify until T1 would wait. 

```
T1: take lock
T1: load local
T1: see that it isn't 42
T2: store(42)
T2: notify_all() // called and locked on a mutex, since held by T1
T1: wait() // release mutex
T2: actually wake CV

```

The same stuff already works in `parallel_alrorithms.cpp`, by the way, even without a comment.

Reply:
>I take it as "T1: see that it is still 42", otherwise we'll just return.

Sorry, yes, I flipped the sense here. Pretend I said T2 stored something other than 42 :)

>locking on a mutex will postpone actual notify until T1 would wait.

It only does that if you are holding the mutex when you call notify_all. Otherwise this is still possible:

```
T2: notify_all() // called and locked on a mutex, since held by T1
```

>The same stuff already works in parallel_alrorithms.cpp, by the way, even without a comment.

Ouch.... uhhh I might have written that bug. But I do observe taking the lock and dropping does prevent the bad case I indicated above so it might work accidentally... let me think about this a bit.

Reply:
>It only does that if you are holding the mutex when you call notify_all

Condition variable takes locked mutex, and it starts waiting for notifications "atomically".
See SleepConditionVariableSRW  documentation:

> Sleeps on the specified condition variable and releases the specified lock as an atomic operation.

Othewise the usual pattern "change condition under lock" would have been broken as well.

> let me think about this a bit.

Take your time. 
I've seen this in as well in https://github.com/ogiroux/atomic_wait , it was very novel idea to me that day ("no-op under lock? what?")